src/MassTransit/Context/InMemoryOutboxMessageSchedulerContext.cs

```diff
@@ -0,0 +1,266 @@
+﻿// Copyright 2007-2016 Chris Patterson, Dru Sellers, Travis Smith, et. al.
+//  
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the 
+// License at 
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0 
+// 
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR 
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the 
+// specific language governing permissions and limitations under the License.
+namespace MassTransit.Context
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using GreenPipes;
+    using Scheduling;
+
+
+    public class InMemoryOutboxMessageSchedulerContext :
+        MessageSchedulerContext
+    {
+        readonly MessageSchedulerContext _context;
+        readonly object _listLock = new object();
+        readonly List<ScheduledMessage> _scheduledMessages;
+
+        public InMemoryOutboxMessageSchedulerContext(MessageSchedulerContext context)
+        {
+            _context = context ?? throw new ArgumentNullException(nameof(context));
+
+            _scheduledMessages = new List<ScheduledMessage>();
+        }
+
+        void AddScheduledMessage(ScheduledMessage scheduledMessage)
+        {
+            lock (_listLock)
+                _scheduledMessages.Add(scheduledMessage);
+        }
+
+        public Task CancelAllScheduledMessages()
+        {
+            ScheduledMessage[] scheduledMessages;
+
+            lock (_listLock)
+                scheduledMessages = _scheduledMessages.ToArray();
+
+            var tasks = new List<Task>();
```

Comment:
`scheduledMessages.Select(scheduledMessage => _context.CancelScheduledSend(scheduledMessage.Destination, scheduledMessage.TokenId)).ToList()` to avoid list resizes

Reply:
agreed, this would be a bit better. I'll make the tweak

Reply:
But can't this result on code within the CancelScheduledSend to happen within the lock, since it executes synchronously until an await is specified?

Reply:
I was honestly going to keep it the same, and just do this:

`var tasks = new List<Task>(scheduledMessages.Length);`

Seems the most straightforward to me.

Reply:
I see it like that, but it's IMHO:
```csharp
public Task CancelAllScheduledMessages()
        {
            Task[] tasks;

            lock (_listLock)
                tasks = _scheduledMessages.Select(scheduledMessage => _context.CancelScheduledSend(scheduledMessage.Destination, scheduledMessage.TokenId)).ToArray()

            if (!tasks.Any())
               return TaskUtils.Completed;
            return Task.WhenAll(tasks);
        }
```

Reply:
Are we just splitting hairs here? Both techniques will do what we want. It's really just coding style at this point.